pkg/auth/server/login/login.go

```diff
@@ -133,19 +158,99 @@ func (l *Login) handleLogin(w http.ResponseWriter, req *http.Request) {
 	l.auth.AuthenticationSucceeded(context, then, w, req)
 }
 
-var DefaultLoginFormRenderer = loginTemplateRenderer{}
+// NewLoginFormRenderer creates a login form renderer that takes in an optional custom template to
+// allow branding of the login page. Uses the default if customLoginTemplateFile is not set.
+func NewLoginFormRenderer(templates *configapi.OAuthTemplates) (*loginTemplateRenderer, error) {
+	r := &loginTemplateRenderer{}
+	if templates != nil && len(templates.Login) > 0 {
+		customTemplate, err := template.ParseFiles(templates.Login)
+		if err != nil {
+			return nil, err
+		}
+		r.loginTemplate = customTemplate
+	} else {
+		r.loginTemplate = defaultLoginTemplate
+	}
+
+	return r, nil
+}
+
+func ValidateLoginTemplate(templates *configapi.OAuthTemplates) fielderrors.ValidationErrorList {
+	allErrs := fielderrors.ValidationErrorList{}
+
+	if templates == nil || len(templates.Login) == 0 {
+		return allErrs
+	}
+
+	if _, err := os.Stat(templates.Login); err != nil {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("templates.login", templates.Login, "could not read file"))
+		// Short circuit here if the file can't be read.
+		return allErrs
+	}
+
+	template, err := template.ParseFiles(templates.Login)
+	if err != nil {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("templates.login", templates.Login, err.Error()))
+	}
+
+	// Execute the template with dummy values and check if they're there.
+	form := LoginForm{
+		Action: "MyAction",
+		Error:  "MyError",
+		Names: LoginFormNames{
+			Then:     "MyThenName",
+			CSRF:     "MyCSRFName",
+			Username: "MyUsernameName",
+			Password: "MyPasswordName",
+		},
+		Values: LoginFormValues{
+			Then:     "MyThenValue",
+			CSRF:     "MyCSRFValue",
+			Username: "MyUsernameValue",
+		},
+	}
+
+	var buffer bytes.Buffer
+	err = template.Execute(&buffer, form)
+	if err != nil {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("templates.login", templates.Login, err.Error()))
+	}
+	output := buffer.Bytes()
+
+	var testFields = map[string]string{
+		"Action":          form.Action,
+		"Error":           form.Error,
+		"Names.Then":      form.Names.Then,
+		"Names.CSRF":      form.Values.CSRF,
+		"Names.Username":  form.Names.Username,
+		"Names.Password":  form.Names.Password,
+		"Values.Then":     form.Values.Then,
+		"Values.CSRF":     form.Values.CSRF,
+		"Values.Username": form.Values.Username,
+	}
+
+	for field, value := range testFields {
+		if !bytes.Contains(output, []byte(value)) {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("templates.login", templates.Login, fmt.Sprintf("template is missing parameter {{ .%s }}", field)))
```

Comment:
e.g. append `fmt.Errorf("template is missing parameter {{ .%s }}", field)` to the list of errors to return, and the caller can use that as a details message in a NewFieldInvalid error